File: bro_code/sorting_algorithms/merge_sort.cpp
#include <iostream>
#include "bits/stdc++.h"


// space complexity - O(n)
// runtime complexity O(n*logn)
// ამ ალგორითმის დროს პირველ რიგში მასივი იშლება სხვა პატარა მასივებად სანამ თითოეულ მასივში არ იარსებებს მხოლოდ ერთი ელემენტი

int* slice(int array[], int start, int end) {
	// მასივის გასაყოფად შექმნილი ფუნქცია
	// ეს ფუნქცია პირდაპირ არ მონაწილეობს ალგორითმში (სხვა ენებში შეიძლება ჩასენებულის გამოყენება)
	int new_arr[] = {};
	int length = sizeof(array)/sizeof(int);
	for(int i=start; i<end; i++) {
		std::cout << array[i];
	}
	return {};
}

void merge_sort(int array[], int length) {

	if(length <= 1) return; // base case, ამ შემთვევაში უკვე აღარაა მასივის გაყოფა საჭირო

	int middle = length / 2;
}

void merge(int leftArr[], int rightArr[], int array[]) {

}

int main() {
	int my_arr[] = {8, 2, 5, 3, 4, 7, 6, 1};
	int *arr[] = {};
	arr[0] = &my_arr[3];
	std::cout << arr[0];
	int length = sizeof(my_arr)/sizeof(int); // array length

	merge_sort(my_arr, length);

	for(int i=0; i<length; i++) {
		std::cout << my_arr[i] << " ";
	}

	return 0;
}

/*
1) მასივის დაშლა
				 [3, 7, 8, 5, 4, 2, 6, 1]
			[3, 7, 8, 5]		 [4, 2, 6, 1]
		 [3, 7]	   [8, 5]	  	[4, 2]   [6, 1]
	   [3] [7]     [8] [5]	   [4] [2]   [6] [1]

2) merge, ამ დაშლილი მასივების თითოეული ელემენტი დარდება ერთმანეთს და ისე იქმნება ახალი მასივი
	1) 3 & 7 --> [3, 7]   
	2) 8 & 5 --> [5, 8]
	[3, 7] & [5, 8]
	3) 3 & 5 --> [3, , ,]
	4) 7 & 5 --> [3, 5, ,]
	5) 7 & 8 --> [3, 5, 7, 8]
	......
			 	[1, 2, 3, 4, 5, 6, 7, 8]
			[3, 5, 7, 8]		 [1, 2, 4, 6]
		 [3, 7]	   [5, 8]	  	[2, 4]   [1, 6]
	   [3] [7]     [8] [5]	   [4] [2]   [6] [1]
*/

File: bro_code/sorting_algorithms/merge_sort.py
# O(n log n) - using more space then bubble/Insertion sort, becouse of sub arraies
# https://youtu.be/CBYHwZcbD-s?t=7918


unsorted_arr = [2, 8, 4, 1, 3]

def merge(leftArr, rightArr, arr):
	print("arraies for merge: ", leftArr, rightArr, arr)
	leftSize = len(arr) // 2
	rightSize = len(arr) - leftSize
	i = l = r = 0  # indexes

	# subarray-is lementebis shedareba da mati dasortirebulad gaertianeba
	# check the conditions for merging-shertskma
	while l < leftSize and r < rightSize:
		# mag 4 (shedareba) 1 --> 4 (shedareba) 3 --> [1, 3] + 4 daemateba bolo while-shi
		if (leftArr[l] < rightArr[r]):
			arr[i] = leftArr[l]
			i += 1
			l += 1
		else:
			arr[i] = rightArr[r]
			i += 1
			r += 1

	#  if there is odd amount of items in any subarray
	while l < leftSize:
		arr[i] = leftArr[l]
		i += 1
		l += 1

	while r < rightSize:
		arr[i] = rightArr[r]
		i += 1
		r += 1

	print("merged array: ", arr)

	return arr



def merge_sort(arr):
	print("array: ", arr)
	length = len(arr)

	# if subarray contains only one item
	if (length <= 1): return  # base case 

	middle = length//2
	leftArr = arr[:middle]
	rightArr = arr[middle:]

	i = 0 
	l = 0  # left array
	j = 0  # right array

	while i < length:
		print("i index element: ", arr[i])
		if (i < middle):
			leftArr[l] = arr[i]
			l += 1
		else:
			rightArr[j] = arr[i]
			j += 1

		i += 1

	# split array, by recursion
	# [2, 8, 4, 1, 3] --> [2, 8] --> [2] & [8] 
	merge_sort(leftArr)  
	# [4, 1, 3] --> [4] & [1, 3] --> [1] & [3]
	merge_sort(rightArr)
	# after this start compare and merge of arrays
	# merge
	mergedArr = merge(leftArr, rightArr, arr)
	# print("sorted: ", mergedArr)
	return mergedArr


print("result: ", merge_sort(unsorted_arr))


# მასივი იშლება საბერეებად რომელებშიც იქნება მხოლოდ ერთი ელემენტი
# კეთდება შედარება და ჯგუფდება ახალის დასორტილი მასივები მათ სრულ შერწყმამდე
